### PR TITLE
Add per team player stats with extra trackers

### PR DESCRIPTION
**Pull request by Pugzy**
Rewrite the stats logic so that statistics can be per team. This allows the stats at the end of the match to show contributions towards both teams.

- Introduces a table to store the team, player, stats values and create new stat entries with a reference to its global stats parent.
- Resets a player's active kills streak when switching teams.
- Store snapshot of player's name in a component inside a teams player stats.
- Rewrite team stats GUI to make use of per-team stats (no longer need to create online/observers/offline players separately.
- Attempts to fix the current issues with the team stats GUI where nicknames and skins are revealed.
- Adds extra stats related to objective completion (rather than just flags).

Still a WIP and requires a bit more testing but general feedback is more than welcome.

Additional changes:

- Tracking activity time for players on a team within the `StatsMatchModule`.
- Ability to check which team was a players primary team (based on how long they played on each team or obs).
- Fixed a PGM issue where any goal change event would trigger not only the invalidation of winners but a check to see if anyone has won. The order of these events meant that a match could be ended before the event of lets say a wool capture would be fired so at the point of match end that wool wouldn't count in a players stats.
- The above should also fix an issue where the announcement of the objective being completed would display after the match over text.
- A slight tweak to the block vector parser so it allows `.5` coordinates that are then translated (this makes structures more backwards compatible as this is the same behaviour as old new PGM).

### CODE REVIEW

File: core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java [hunk 1]
```diff
@@ -97,8 +107,12 @@ public StatsMatchModule(Match match) {
...
   }
 
-  public Map<UUID, PlayerStats> getStats() {
-    return Collections.unmodifiableMap(allPlayerStats);
+  public Map<UUID, PlayerStats> getGlobalStats() {
+    return allPlayerStats;
+  }
+
+  public Table<Team, UUID, PlayerStats> getStats() {
+    return Tables.unmodifiableTable(stats);
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java]: keep getStats signature intact (returning global stats) and have getParticipationStats as the new one returning a table, for api backwards-compatibility

File: core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java [hunk 2]
```diff
@@ -138,13 +152,57 @@ public void onDestroyableBreak(DestroyableHealthChangeEvent event) {
...
+
+  @EventHandler(priority = EventPriority.MONITOR)
+  public void onCoreLeak(CoreLeakEvent event) {
+    event
+        .getCore()
+        .getContributions()
+        .forEach(
+            leaker -> {
+              if (leaker.getPlayerState() != null) {
+                // TODO: wrong method
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java]: ?

File: core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java [hunk 3]
```diff
@@ -138,13 +152,57 @@ public void onDestroyableBreak(DestroyableHealthChangeEvent event) {
...
+    }
   }
 
   @EventHandler(priority = EventPriority.MONITOR)
-  public void onFlagHold(FlagPickupEvent event) {
-    getPlayerStat(event.getCarrier()).onFlagPickup();
+  public void onWoolCapture(PlayerWoolPlaceEvent event) {
+    if (event.getPlayer() != null && event.getPlayer().getPlayer().isPresent()) {
+      getPlayerStat(event.getPlayer()).onWoolCapture();
+    }
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/stats/StatsMatchModule.java]: is this about getting the player off of a player state? If so, that's not needed, keep it as:
```suggestion
    if (event.getPlayer() != null) {
      getPlayerStat(event.getPlayer()).onWoolCapture();
    }
```

**Review by Pablete1234 — CHANGES_REQUESTED** [review]

File: core/src/main/java/tc/oc/pgm/stats/PlayerStats.java [hunk 4]
```diff
@@ -39,17 +55,33 @@ public class PlayerStats {
   // See StatsMatchModule#sendLongHotbarMessage
   private Future<?> hotbarTaskCache;
 
+  public PlayerStats() {
+    this.parent = null;
+    this.component = null;
+    this.timePlayed = Duration.ZERO;
+    this.inTime = null;
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/stats/PlayerStats.java]:
```suggestion
    this(null, null);
```

File: core/src/main/java/tc/oc/pgm/stats/PlayerStats.java [hunk 5]
```diff
@@ -187,4 +274,32 @@ public Future<?> getHotbarTask() {
...
+    if (parent != null) parent.endParticipation();
+  }
+
+  public Duration getTimePlayed() {
+    // If not ended yet add the current session time up
+    return inTime == null ? timePlayed : timePlayed.plus(getActiveSessionDuration());
+  }
+
+  public Duration getActiveSessionDuration() {
+    return (inTime == null) ? Duration.ZERO : Duration.between(inTime, Instant.now());
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/stats/PlayerStats.java:300]:
```suggestion
    return inTime == null ? Duration.ZERO : Duration.between(inTime, Instant.now());
```

File: core/src/main/java/tc/oc/pgm/stats/menu/items/TeamStatsMenuItem.java [hunk 6]
```diff
@@ -40,43 +42,28 @@ public class TeamStatsMenuItem implements MenuItem {
...
-      Competitor team,
-      Collection<MatchPlayer> relevantObservers,
-      Collection<UUID> relevantOfflinePlayers) {
-    StatsMatchModule smm = match.needModule(StatsMatchModule.class);
+  public TeamStatsMenuItem(Match match, Competitor team, Map<UUID, PlayerStats> playerStats) {
 
     this.team = team;
     this.members = Lists.newArrayList();
-    this.stats = new TeamStats(team, smm);
+    this.stats = new TeamStats(team, new ArrayList<>(playerStats.values()));
```
Comment (Pablete1234) [inline core/src/main/java/tc/oc/pgm/stats/menu/items/TeamStatsMenuItem.java]: i'd avoid the relocation, just pass accept a Collection<PlayerStats> in the constructor instead
```suggestion
    this.stats = new TeamStats(team, playerStats.values());
```

**Review by Pablete1234 — COMMENTED** [review]
just a few tweaks

**Review by Pablete1234 — APPROVED** [review]

File: core/src/main/java/tc/oc/pgm/goals/GoalMatchModule.java [hunk 7]
```diff
@@ -147,7 +147,9 @@ protected void updateProgress(Goal goal) {
     for (Competitor competitor : competitorsByGoal.get(goal)) {
       progressByCompetitor.put(competitor, new GoalProgress(competitor));
     }
-    match.calculateVictory();
+
+    // Forces team rankings to be invalidated
+    match.getWinners();
```
Comment (Electroid) [inline core/src/main/java/tc/oc/pgm/goals/GoalMatchModule.java:152]: This is subtle behavior which is not great, it’s better if we could do something explicit. But I’ll still merge for now.

**Review by Electroid — APPROVED** [review]

**Pull request merged**